src/text-input/src/TextInput.js

```diff
@@ -68,14 +87,24 @@ class TextInput extends PureComponent {
     spellCheck: true
   }
 
+  getValue() {
+    if ('value' in this.props) {
+      // Console.log('Input has controlled value prop')
+      return this.props.value
+    }
+    return this.state.value
```

Comment:
To simplify things like state vs props we might benefit from making two separate (both exported) classes. One `*Stateless` that's fully controlled, and another that's stateful. That way consumers can decide which one to use. What do you think?

I'm happy to provide a bit more feedback on this if you need a little more suggestion here.

Reply:
I love this idea, appreciate the guidance. I'll see what I can put together!

Reply:
I should also follow up that we could use the `*Stateless` version _within_ the stateful one.

Reply:
Can you give me a clue on the file structure this might impact? Where do you want each component ... Exact filenames and exports in some fashion I can follow would be appreciated.

Reply:
I have done some legwork and have it working just would rather ask ahead of pushing

Reply:
I imagine the stateful component (`TextInput`) would look something like this. Note: I'm excluding a lot of the bits that you've added, but anything that's managing state would belong in this component instead of `StatelessTextInput`. Atlaskit actually does this quite a bit, iirc.

```jsx
import React, { PureComponent } from 'react'
import StatelessTextInput from './StatelessTextInput'

class TextInput extends PureComponent {
  static propTypes = {
    /**
     * Composes the StatelessTextInput component as the base.
     */
    ...StatelessTextInput.propTypes
  }

  state = {
    value: this.props.value
  }

  handleChange = (event) => {
    // handle state change and bubble event to `this.props.onChange` if provided?
  }

  render() {
    return (
      <TextStatelessTextInput
        {...this.props}
        value={this.state.value}
        onChange={this.handleChange}
      />
    )
  }
}

export default TextInput
```

Sidenote: I'm also beginning to wonder if this behavior will interfere with native browser support for clearing inputs (IE / Edge)?

Reply:
Okay great. This is generally what I have already, and your naming preferences help a ton! 

ON the IE / Edge *edge* case, it seems to be the accepted practice to just `display:none` on it. 
https://stackoverflow.com/questions/14007655/remove-ie10s-clear-field-x-button-on-certain-inputs
https://developer.mozilla.org/en-US/docs/Web/CSS/::-ms-clear

I was thinking of just adding a style to the components with a prop. Do you think it should have a true/false that can be set by consumer of the component? ..or is it good enough to simply document that we override it completely since it is non-standard?

Reply:
It's probably fine for us to override it with a cross-browser version only if they've opted in to using `isClearable`, where `isClearable` defaults to `false`. That way we don't introduce a breaking change for consumers today, and if they prefer the native non-cross-browser behavior they can continue to us it.